src/claims.rs

```diff
@@ -0,0 +1,456 @@
+use crate::scope::{Scope, ScopeValue};
+use serde::{de::DeserializeOwned, Deserialize, Deserializer, Serialize};
+use serde_with::skip_serializing_none;
+
+/// Functions as the `claims` parameter inside a [`crate::SiopRequest`].
+#[derive(Default, Debug, Clone, PartialEq, Deserialize, Serialize)]
+pub struct ClaimRequests {
+    pub user_claims: Option<StandardClaims<IndividualClaimRequest>>,
+    pub id_token: Option<StandardClaims<IndividualClaimRequest>>,
+}
+
+/// [`Claim`] trait that is implemented by both [`ClaimValue`] and [`IndividualClaimRequest`].
+pub trait Claim: Default + Clone + DeserializeOwned {
+    type ValueType;
+    type ClaimType<S>: Claim<ValueType = S> + Serialize
+    where
+        S: Serialize + Default + Clone + DeserializeOwned;
+}
```

Comment:
Reordering the bounds a little bit makes it more easy to understand for me and seems equivalent. 

```suggestion
pub trait Claim: Default + Clone + DeserializeOwned + Serialize {
    type ValueType;
    type ClaimType<S>: Claim<ValueType = S>
    where
        S: Default + Clone + DeserializeOwned + Serialize;
}
```

The more I think about it, the less  I think it's possible to remove one associated type, since the code needs to say something (i.e. it's serializable) about the wrapper (self or the type implementing claim), and the inside what it's wrapping (i.e. also serializable). That's basically all the trait is saying right? That's not so much, makes me wonder even more if the trait should be there in the first place, or if a more explicit enum based variant would uphold all guarantees while being less elegant but also way simpler.

Reply:
After much trial and error I managed to get it down to one single associative type and reduce the amount of trait bounds. In the core it looks like this:
```
mod sealed {
    /// [`Claim`] trait that is implemented by both [`ClaimValue`] and [`ClaimRequest`].
    pub trait Claim {
        type Container<T>;
    }
}

#[derive(Default, Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct ClaimValue<T = ()>(pub T);

impl<T> sealed::Claim for ClaimValue<T> {
    type Container<U> = U;
}

#[derive(Default, Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct ClaimRequest<T = ()>(IndividualClaimRequest<T>);

impl<T> sealed::Claim for ClaimRequest<T> {
    type Container<U> = IndividualClaimRequest<U>;
}

#[derive(Default, Debug, Clone, PartialEq, Deserialize, Serialize)]
#[serde(untagged)]
pub enum IndividualClaimRequest<T> {
    #[default]
    Null,
    Object {
        #[serde(skip_serializing_if = "Option::is_none")]
        essential: Option<bool>,
        #[serde(skip_serializing_if = "Option::is_none")]
        value: Option<T>,
        #[serde(skip_serializing_if = "Option::is_none")]
        values: Option<Vec<T>>,
        #[serde(flatten, deserialize_with = "parse_other")]
        other: Option<serde_json::Value>,
    },
}

#[skip_serializing_none]
#[derive(Debug, Clone, Default, PartialEq, Serialize, Deserialize)]
#[serde(default, deny_unknown_fields)]
pub struct StandardClaims<C: sealed::Claim> {
    #[serde(bound(serialize = "C::Container<String>: Serialize"))]
    #[serde(bound(deserialize = "C::Container<String>: Deserialize<'de> + Default"))]
    #[serde(deserialize_with = "parse_optional_claim")]
    pub name: Option<C::Container<String>>,
    #[serde(bound(serialize = "C::Container<String>: Serialize"))]
    #[serde(bound(deserialize = "C::Container<String>: Deserialize<'de> + Default"))]
    #[serde(deserialize_with = "parse_optional_claim")]
    pub family_name: Option<C::Container<String>>,
}
```

So imo pretty clean and to the point. This approach guarantees the typesafety of the `StandardClaims` struct for both variations.

Reply:
> That's basically all the trait is saying right? That's not so much, makes me wonder even more if the trait should be there in the first place, or if a more explicit enum based variant would uphold all guarantees while being less elegant but also way simpler.
The main purpose of the trait is to guarantee object type safety of the StandardClaims struct. The `Default + Clone + DeserializeOwned + Serialize` trait bounds admittedly was more a result of me fighting with the compiler. In this new implementation the functionality is reduced to exactly that what it is supposed to do, i.e. guaranteeing object type safety for `StandardClaims`. 

The added benefit of this is that wrong usage of the standard claims can be caught at compile time instead of runtime.